// eslint-plugin-weavers-react/constants.js
module.exports.RULES = {
  REACT_MAXLINES_PER_FUNCTION: 'react_maxlines_per_function',
};

module.exports.NODE_TYPES = {
  FUNCTION_EXPESSION: 'FunctionExpression',
  FUNCTION_DECLARATION: 'FunctionDeclaration',
  ARRROW_FUNCTION_EXPRESSION: 'ArrowFunctionExpression',
  JSX_ELEMENT: 'JSXElement',
};

module.exports.METADATA = {
  TYPE: 'suggestion',
  DOCUMENT: {
    description: 'Enforce a maximum number or lines in functions',
    category: 'Stylistic Issues',
    recommended: true,
    url: '',
  },
  SCHEMA: {
    type: 'object',
    properties: {
      maxJSLines: {
        type: 'integer',
        minimum: 0,
      },
      maxJSXLines: {
        type: 'integer',
        minimum: 0,
      },
      maxLines: {
        type: 'integer',
        minimum: 0,
      },
    },
    additionalProperties: false,
  },
  MESSAGES: {
    maxLines:
      'Function {{functionName}} exceeds maximum lines({{totalLinesInNode}}).Maximum allowed is {{maxLines}}.',
    maxJSLines:
      'Function {{functionName}} exceeds maximum Javascript lines({{totalJSLines}}). Maximum allowed is {{maxJSLines}}.',
    maxJSXLines:
      'Function {{functionName}} exceeds maximum JSX lines({{totalJSXLines}}). Maximum allowed is {{maxJSXLines}}.',
  },
};

module.exports.CONST = {
  RENDER: 'render',
  NUMBER: 'number',
  MAX_LINES_DEFAULT: 80,
  MAX_JS_LINES_DEFAULT: 30,
  MAX_JSX_LINES_DEFAULT: 50,
  END_LINE: '\n',
};
